File: summary/elixir/7_greatest_common_divisor.exs
defmodule MathUtils do
  # ユークリッドの互除法による最大公約数を求める関数
  def gcd(a, 0), do: a
  def gcd(a, b), do: gcd(b, rem(a, b))
end

IO.puts(MathUtils.gcd(48, 18)) # => 6

File: summary/ts/7_greatest_common_divisor.ts
function gcd(a: number, b: number): number {
  // ベースケース: bが0の場合、aが最大公約数
  if (b === 0) {
    return a;
  }

  // aをbで割った余りをcとする
  const c = a % b;

  // bとcの最大公約数を再帰的に求める
  return gcd(b, c);
}

console.log(gcd(24, 18)); // 6
console.log(gcd(30, 20)); // 10
